### DISCUSSION TITLE
Best practice for Atomic Design Systems

### DISCUSSION THREAD

**Discussion by thomasfarre** [tailwindlabs/tailwindcss#15195] [Help]
Hi everyone, I hope you're doing well! 👋

I'm working on implementing an Atomic Design System using React, TailwindCSS, and Storybook, and I'm looking for some suggestions on the best way to organize things. Here's the current setup:

1. **Atoms**: I’ve defined basic atomic components (e.g., colors, typography, spacing) as Storybook stories, using TailwindCSS properties to represent their style and behavior. For example:

![Screenshot from 2024-11-26 11-20-09](https://github.com/user-attachments/assets/80294c99-e94d-49b5-9eed-0f1b3ec0cece)

2. **Molecules**: Components like buttons or cards combine these atoms. However, I’ve hit a challenge:
        
- If I want to update a TailwindCSS class used in multiple molecules (e.g., change the base padding or typography), I’d have to update each molecule manually.

- This doesn’t feel scalable or efficient.

### My Solutions So Far

To address this, I’ve considered a couple of approaches:

1. **Atom as React Components**: Define a React component for each atom (e.g., PaddingAtom, ColorAtom) and reuse these in molecules. However, this feels overly complex and not very lightweight for something as simple as styles.

2. **Global CSS Variables**: Define global CSS variables for atomic properties (e.g., colors, spacings) in a @layer in TailwindCSS like this:

```
    @layer base {
        :root {
            --color-primary: theme('colors.blue.500');
            --color-secondary: theme('colors.green.500');
            --padding-default: theme('spacing.4');
        }
    }
```

Then reference these variables in Tailwind utility classes, ensuring consistency across molecules and making updates easier. It seems similar to what TailwindCSS is pushing for with v4.0.

![Screenshot from 2024-11-26 11-09-20](https://github.com/user-attachments/assets/793ab6da-333a-465c-b577-59612708fcfc)

### My Questions

1. Is defining global CSS variables for atomic properties considered a good practice in TailwindCSS?

2. Would this approach make maintaining a design system easier, especially when working on larger projects or scaling to new ones?

3. Are there better ways to combine Atomic Design principles with TailwindCSS and React without adding too much complexity?

I’d love to hear your thoughts or suggestions! Thanks in advance for your help! 🙌

**Comment by denis23x** [comment]
Hi @thomasfarre 

You did a great work 👍

I really liked this https://sass-guidelin.es implementation, I never encountered any problems, they call it 7-1 pattern.
